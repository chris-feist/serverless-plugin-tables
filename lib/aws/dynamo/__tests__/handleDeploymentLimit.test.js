
const handleDeploymentLimit = require('../handleDeploymentLimit');

const { createTableResource } = require('../testUtils');

beforeEach(() => {
  jest.clearAllMocks();
});

describe('handleDeploymentLimit', () => {
  describe('checkDependency', () => {
    test('finds dependency', () => {
      const dependsOn = ['dep1', 'dep2', 'dep3', 'dep4'];
      const remainingTableResources = new Set(['depA', 'depB', 'dep3', 'depD']);

      const result = handleDeploymentLimit.checkDependency(dependsOn, remainingTableResources);

      expect(result).toEqual(true);
    });

    test('no matching dependencies', () => {
      const dependsOn = ['dep1', 'dep2', 'dep3', 'dep4'];
      const remainingTableResources = new Set(['depA', 'depB', 'depC', 'depD']);

      const result = handleDeploymentLimit.checkDependency(dependsOn, remainingTableResources);

      expect(result).toEqual(false);
    });
  });

  describe('createBatch', () => {
    test('creates a batch with internal dependencies', () => {
      const remainingTables = [
        createTableResource({ resourceName: 'res1', dependsOn: 'res9' }),
        createTableResource({ resourceName: 'res2', dependsOn: 'res1000' }),
        createTableResource({ resourceName: 'res3' }),
        createTableResource({ resourceName: 'res4' }),
        createTableResource({ resourceName: 'res5' }),
        createTableResource({ resourceName: 'res6' }),
        createTableResource({ resourceName: 'res7' }),
        createTableResource({ resourceName: 'res8' }),
        createTableResource({ resourceName: 'res9' }),
        createTableResource({ resourceName: 'res10', dependsOn: 'res9' }),
        createTableResource({ resourceName: 'res11', dependsOn: 'res10' }),
        createTableResource({ resourceName: 'res12', dependsOn: 'res11' }),
      ];
      const batchSize = 5;

      const result = handleDeploymentLimit.createBatch(remainingTables, batchSize);

      expect(result).toMatchSnapshot();
      expect(remainingTables).toMatchSnapshot();
    });

    test('creates a batch assigns depends on', () => {
      const remainingTables = [
        createTableResource({ resourceName: 'res1', dependsOn: 'res4' }),
        createTableResource({ resourceName: 'res2', dependsOn: 'res5' }),
        createTableResource({ resourceName: 'res3' }),
        createTableResource({ resourceName: 'res4' }),
        createTableResource({ resourceName: 'res5', dependsOn: 'res999' }),
      ];
      const batchSize = 5;
      const dependsOn = 'otherRes';

      const result = handleDeploymentLimit.createBatch(remainingTables, batchSize, dependsOn);

      expect(Object.values(result[0])[0].DependsOn[0]).toEqual(dependsOn);
      expect(result).toMatchSnapshot();
      expect(remainingTables).toMatchSnapshot();
    });
  });

  test('batches tables', () => {
    const plugin = {
      debug: jest.fn(),
      getOptions: jest.fn(() => ({})),
    };
    const tables = []; // ignored
    const incomingTables = [
      createTableResource({ resourceName: 'res1', dependsOn: 'res9' }),
      createTableResource({ resourceName: 'res2', dependsOn: 'res1' }),
      createTableResource({ resourceName: 'res3' }),
      createTableResource({ resourceName: 'res4' }),
      createTableResource({ resourceName: 'res5' }),
      createTableResource({ resourceName: 'res6' }),
      createTableResource({ resourceName: 'res7' }),
      createTableResource({ resourceName: 'res8' }),
      createTableResource({ resourceName: 'res9' }),
      createTableResource({ resourceName: 'res10', dependsOn: 'res9' }),
      createTableResource({ resourceName: 'res11', dependsOn: 'res10' }),
      createTableResource({ resourceName: 'res12', dependsOn: 'res11' }),
    ];

    const result = handleDeploymentLimit(plugin, tables, incomingTables);

    expect(result).toMatchSnapshot();
  });

  test('circular dependency', () => {
    const plugin = {
      debug: jest.fn(),
      getOptions: jest.fn(() => ({})),
    };
    const tables = []; // ignored
    const incomingTables = [
      createTableResource({ resourceName: 'res1', dependsOn: 'res3' }),
      createTableResource({ resourceName: 'res2', dependsOn: 'res1' }),
      createTableResource({ resourceName: 'res3', dependsOn: 'res2' }),
    ];

    expect(() => handleDeploymentLimit(plugin, tables, incomingTables)).toThrow();
  });
});
