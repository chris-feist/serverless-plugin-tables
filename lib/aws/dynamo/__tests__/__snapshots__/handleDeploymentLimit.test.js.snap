// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleDeploymentLimit batches tables 1`] = `
Array [
  Object {
    "res3": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res4": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res5": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res6": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res7": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res8": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res9": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res1": Object {
      "DependsOn": Array [
        "res9",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res10": Object {
      "DependsOn": Array [
        "res9",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res2": Object {
      "DependsOn": Array [
        "res1",
        "res10",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res11": Object {
      "DependsOn": Array [
        "res10",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res12": Object {
      "DependsOn": Array [
        "res11",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
]
`;

exports[`handleDeploymentLimit createBatch creates a batch assigns depends on 1`] = `
Array [
  Object {
    "res3": Object {
      "DependsOn": Array [
        "otherRes",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res4": Object {
      "DependsOn": Array [
        "otherRes",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res5": Object {
      "DependsOn": Array [
        "res999",
        "otherRes",
      ],
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
]
`;

exports[`handleDeploymentLimit createBatch creates a batch assigns depends on 2`] = `
Array [
  Object {
    "res1": Object {
      "DependsOn": "res4",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res2": Object {
      "DependsOn": "res5",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
]
`;

exports[`handleDeploymentLimit createBatch creates a batch with internal dependencies 1`] = `
Array [
  Object {
    "res2": Object {
      "DependsOn": "res1000",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res3": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res4": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res5": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res6": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
]
`;

exports[`handleDeploymentLimit createBatch creates a batch with internal dependencies 2`] = `
Array [
  Object {
    "res1": Object {
      "DependsOn": "res9",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res7": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res8": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res9": Object {
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res10": Object {
      "DependsOn": "res9",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res11": Object {
      "DependsOn": "res10",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
  Object {
    "res12": Object {
      "DependsOn": "res11",
      "Properties": Object {
        "TableName": "TestTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
]
`;
